x/ibc/testing/chain.go

```diff
@@ -0,0 +1,373 @@
+//
+// The first block height is committed to state in order to allow for client creations on counterparty
+// chains. The TestChain will return with a block height starting at 2.
+//
+// Time management is handled by the IBCTestSuite in order to ensure synchrony between chains. Each
+// update of any chain increments the block header time by 5 seconds.
+func NewTestChain(t *testing.T, clientID string) *TestChain {
+	// generate validator private/public key
+	privVal := tmtypes.NewMockPV()
+	pubKey, err := privVal.GetPubKey()
+	require.NoError(t, err)
+
+	// create validator set with single validator
+	validator := tmtypes.NewValidator(pubKey, 1)
+	valSet := tmtypes.NewValidatorSet([]*tmtypes.Validator{validator})
+	signers := []tmtypes.PrivValidator{privVal}
+
+	app := simapp.Setup(false)
+	ctx := app.BaseApp.NewContext(false,
+		abci.Header{
+			Height: 1,
+			Time:   globalStartTime,
+		},
+	)
+
+	// generate and set SenderAccount
+	senderPrivKey := secp256k1.GenPrivKey()
+	simapp.AddTestAddrsFromPubKeys(app, ctx, []crypto.PubKey{senderPrivKey.PubKey()}, sdk.NewInt(10000000000))
+	acc := app.AccountKeeper.GetAccount(ctx, sdk.AccAddress(senderPrivKey.PubKey().Address()))
+
+	// commit init chain changes so create client can be called by a counterparty chain
+	app.Commit()
+	// create current header and call begin block
+	header := abci.Header{
+		Height: 2,
+		Time:   globalStartTime.Add(timeIncrement),
+	}
+	app.BeginBlock(abci.RequestBeginBlock{Header: header})
+
+	lastHeader := ibctmtypes.CreateTestHeader(clientID, 1, globalStartTime, valSet, signers)
+
+	// create an account to send transactions from
+	return &TestChain{
+		t:             t,
+		ClientID:      clientID,
+		App:           app,
+		LastHeader:    lastHeader,
+		CurrentHeader: header,
+		Querier:       keeper.NewQuerier(*app.IBCKeeper),
+		Vals:          valSet,
+		Signers:       signers,
+		senderPrivKey: senderPrivKey,
+		SenderAccount: acc,
+	}
+}
+
+// GetContext returns the current context for the application.
+func (chain *TestChain) GetContext() sdk.Context {
+	return chain.App.BaseApp.NewContext(false, chain.CurrentHeader)
+}
+
+// QueryProof performs an abci query with the given key and returns the merkle proof for the query
+// and the height at which the query was performed.
+func (chain *TestChain) QueryProof(key []byte) (commitmenttypes.MerkleProof, uint64) {
+	res := chain.App.Query(abci.RequestQuery{
+		Path:   fmt.Sprintf("store/%s/key", host.StoreKey),
+		Height: chain.App.LastBlockHeight(),
+		Data:   key,
+		Prove:  true,
+	})
+
+	proof := commitmenttypes.MerkleProof{
+		Proof: res.Proof,
+	}
+
+	return proof, uint64(res.Height)
+}
+
+// nextBlock sets the last header to the current header and increments the current header to be
+// at the next block height. It does not update the time as that is handled by the IBCTestSuite.
+//
+// CONTRACT: this function must only be called after app.Commit() occurs
+func nextBlock(chain *TestChain) {
```

Comment:
I'd make this public so that other SDK apps can use it:

```suggestion
func (chain *TestChain) NextBlock() {
```